utopia-remix/app/handlers/validators.ts

```diff
@@ -0,0 +1,37 @@
+import { AccessValidator, ensure, getUser } from '../util/api.server'
+import { UserProjectPermission } from '../types'
+import { Params } from '@remix-run/react'
+import { Status } from '../util/statusCodes'
+import { hasUserProjectPermission } from '../services/permissionsService.server'
+import { getProjectById } from '../models/project.server'
+
+export function validateProjectAccess(
+  permission: UserProjectPermission,
+  {
+    errorMessage,
+    status,
+  }: {
+    errorMessage?: string
+    status?: number
+  } = {},
+): AccessValidator {
+  return async function (req: Request, params: Params<string>) {
+    const { id: projectId } = params
+    ensure(projectId != null, 'project id is null', Status.BAD_REQUEST)
+
+    const projectData = await getProjectById({ id: projectId })
+    ensure(projectData != null, 'Project not found', Status.NOT_FOUND)
+
+    let user = await getUser(req)
+    let userId = user?.user_id ?? 'ANON'
```

Comment:
I really dislike this use of a hardcoded string here for an anonymous user. Since we're only using it in one place it's fine, but I'd argue that if we have to do this then we should probably provide the fallback string inside `checkUserProjectPermission` rather than inside the validator

Reply:
Yeah, we need it since Okta needs some kind of string here, but I agree that inside the check function might be a better idea. Will change.